design/keycloak.x/storage-architecture.md

```diff
@@ -0,0 +1,129 @@
+often in a version-incompatible way, upon start of a new version means
+requirement for stop-the-world updates and gives little chance for no
+downtime upon upgrade.
+
+*Design-wise*, part of the problem is that the interfaces for accessing
+data objects are intertwined. For example the realm provider is used to
+access not only realm data, but also e.g. clients and roles. The flaw of
+this approach can be seen e.g. in the case of `ClientProvider` which has
+a method `addRole` for adding a client role but no method to remove one -
+`removeRole` is only found in `RealmProvider`. This flaw stems from the
+expectation that the data store is (apart from the user and session
+provider) bound to use the same physical store.
+
+## Logical view
+
+There was always a dichotomy between what is considered cached and what
+stored, especially with the user sessions. “Usual” objects are *stored*
+in a database, yet online sessions are *stored* in Infinispan cache while
+the other objects are *cached* there.
+
+To address this, the split of the storages for the different objects will
+happen - each of the first-class objects like realms or clients (see
+below) will be stored and cached in generally a different store. That
+way, the user will be able to tune the system store and caching engines
+to have e.g. realm definitions stored in files and cached in Infinispan,
+clients stored in etcd store and cached in plain hashmaps and sessions
+shared in Infinispan caches. (This is indeed overstretched but you get
+the picture.) With this split it comes that the references to other
+objects would be loose, and it will be Keycloak responsibility to keep
+referential integrity.
+
+### First-class objects
+
+First-class objects are objects that have a common store settings, and
+their separation is logical  from the perspective of objects handled in
+Keycloak. The first-class objects are depicted below in the color boxes.
+A **realm** contains all data related to the realm that almost never
+changes (realm name and other attributes, identity providers, user
+federation, flows setup of keys etc.). A **client** contains similarly
+all the data related to the client, and **client scope** description of
+the client scope attachable to a client. **Authorization** store contains
+definitions of resources, scopes, policies and permissions. **User**,
+**role**, and **group**, as well as **event** are self-explanatory.
+Sessions have data stored separately for **authentication sessions** and
+**user sessions**.
+
+![First-class objects](img/storage-first-class-objects.png)
+
+### Store configuration
+
+For each object type, a separate store can be configured. The store
+allows configuring persistence and caching independently. For the current
+store, JPA would be used for persistence and Infinispan for caching in
+case of realms, roles etc., and Infinispan would be used for persistence
+and passthrough (i.e. nothing) for caching in case of sessions.
```

Comment:
What about cross-site replication? I don't see it mentioned anywhere, yet it needs to somehow fit into the picture. 

Since we have the caching and the storage layer, where would you plug the x-site in?

Reply:
Both will be affected - caching at least by invalidation events, storage by defining sharing / sharding policy.